runtime/vm/xr64/UpcallThunkGen.cpp

```diff
@@ -31,36 +31,662 @@
+#define STRB(rt, rn, imm12)    (0x39000000 | (rt) | ((rn) << 5) | (((imm12) & 0x00000fff) << 10)) // imm12 must be >= 0
+#define VLDRS(rt, rn, imm12)   (0xBD400000 | (rt) | ((rn) << 5) | (((imm12) & 0x00003ffc) << 8)) // imm12 must be >= 0, multiple of 4
+#define VSTRS(rt, rn, imm12)   (0xBD000000 | (rt) | ((rn) << 5) | (((imm12) & 0x00003ffc) << 8)) // imm12 must be >= 0, multiple of 4
+#define VLDRD(rt, rn, imm12)   (0xFD400000 | (rt) | ((rn) << 5) | (((imm12) & 0x00007ff8) << 7)) // imm12 must be >= 0, multiple of 8
+#define VSTRD(rt, rn, imm12)   (0xFD000000 | (rt) | ((rn) << 5) | (((imm12) & 0x00007ff8) << 7)) // imm12 must be >= 0, multiple of 8
+#define ADD(rd, rn, imm12)     (0x91000000 | (rd) | ((rn) << 5) | ((imm12) << 10)) // ADD (immediate) GPR, imm12 must be >= 0
+#define SUB(rd, rn, imm12)     (0xD1000000 | (rd) | ((rn) << 5) | ((imm12) << 10)) // SUB (immediate) GPR, imm12 must be >= 0
+#define MOV(rd, rm)            (0xaa0003e0 | (rd) | ((rm) << 16)) // MOV (register)
+#define MOVZ(rd, imm16)        (0xd2800000 | (rd) | ((imm16) << 5)) // MOVZ
+#define MOVK(rd, imm16, lsl)   (0xf2800000 | (rd) | ((imm16) << 5) | ((lsl) << 21) ) // MOVK
+#define LSL16 1
+#define LSL32 2
+#define LSL48 3
+#define CBNZ(rt, imm19)        (0xB5000000 | (rt) | (((imm19) & 0x1ffffc) << 3)) // imm19 must be multiple of 4
+#define BR(rn)                 (0xD61F0000 | ((rn) << 5))
+#define BLR(rn)                (0xD63F0000 | ((rn) << 5)) // branch and link
+#define RET(rn)                (0xD65F0000 | ((rn) << 5))
+#define C_SP 31 // x31
+#define LR 30 // x30
+#define HIDDEN_PARAM_REG 8 // x8
+
+#define ROUND_UP_SLOT(si)      (((si) + 7) / 8)
+#define ROUND_UP_SLOT_BYTES(si) ((ROUND_UP_SLOT(si)) * 8)
+
+#define MAX_PARAM_GPR_NUM 8
+#define MAX_PARAM_FPR_NUM 8
+
+#define MAX_BYTES_COPY_BACK_STRAIGHT 64
+
 /**
- * @brief Generate the appropriate thunk/adaptor for a given J9UpcallMetaData
+ * @brief Generate straight sequence of instructions to copy back result
+ * @param instrArray[in/out] A pointer to the thunk memory
+ * @param currIdx[in/out] A pointer to the current instruction index
+ * @param resSize[in] The size in byte to copy, guarantee to be not more than 64
+ * @param paramOffset[in] Offset to the parameter area
+ * @return none
  *
- * @param metaData[in/out] a pointer to the given J9UpcallMetaData
- * @return the address for this future upcall function handle, either the thunk or the thunk-descriptor
+ * Details:
+ *   a static routine to generate instructions to copy back the upcall result
+ *   fixed registers are used
+ *     load the hidden parameter into register x1
+ *     load the result in sequence in no more than 8 registers starting from register x2
+ *     store these registers back into the memory designated by the hidden parameter
+ */
+static void
+copyBackStraight(I_32 *instrArray, I_32 *currIdx, I_32 resSize, I_32 paramOffset)
+{
+	I_32 localIdx = *currIdx;
+	I_32 roundUpSlots = ROUND_UP_SLOT(resSize);
```

Comment:
Given the fact that OSX is not using the stack slot by slot, can you take this convenience of rounding up the size? For example, in case, this copy can overdo it, trashing the last slot.  Power can do it this way because stack is always used slot by slot ... i have never seen a slot shared by more than one entity.

Reply:
The memory area for the return value pointed by the hidden parameter is guaranteed to has the size multiple of 8 bytes.
You can safely copy the return value using the LDR-STR sequence.